chain/index/gc.go

```diff
@@ -0,0 +1,99 @@
+package index
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/lotus/chain/actors/builtin"
+)
+
+var (
+	log             = logging.Logger("chainindex")
+	cleanupInterval = time.Duration(4) * time.Hour
+)
+
+func (si *SqliteIndexer) gcLoop() {
+	defer si.wg.Done()
+
+	// Initial cleanup before entering the loop
+	si.gc(si.ctx)
+
+	cleanupTicker := time.NewTicker(cleanupInterval)
+	defer cleanupTicker.Stop()
+
+	for si.ctx.Err() == nil {
+		si.closeLk.RLock()
+		if si.closed {
+			si.closeLk.RUnlock()
+			return
+		}
+		si.closeLk.RUnlock()
+
+		select {
+		case <-cleanupTicker.C:
+			si.gc(si.ctx)
+		case <-si.ctx.Done():
+			return
+		}
+	}
+}
+
+func (si *SqliteIndexer) gc(ctx context.Context) {
+	if si.gcRetentionEpochs <= 0 {
+		log.Info("gc retention epochs is not set, skipping gc")
+		return
+	}
+	log.Info("starting index gc")
+
+	head := si.cs.GetHeaviestTipSet()
+
+	removalEpoch := int64(head.Height()) - si.gcRetentionEpochs - 10 // 10 is for some grace period
+	if removalEpoch <= 0 {
+		log.Info("no tipsets to gc")
+		return
+	}
+
+	log.Infof("gc'ing all (reverted and non-reverted) tipsets before epoch %d", removalEpoch)
+
+	res, err := si.stmts.removeTipsetsBeforeHeightStmt.ExecContext(ctx, removalEpoch)
+	if err != nil {
+		log.Errorw("failed to remove reverted tipsets before height", "height", removalEpoch, "error", err)
+		return
+	}
+
+	rows, err := res.RowsAffected()
+	if err != nil {
+		log.Errorw("failed to get rows affected", "error", err)
+		return
+	}
+
+	log.Infof("gc'd %d tipsets before epoch %d", rows, removalEpoch)
+
+	// -------------------------------------------------------------------------------------------------
+	// Also GC eth hashes
+
+	// Convert gcRetentionEpochs to number of days
+	gcRetentionDays := si.gcRetentionEpochs / (builtin.EpochsInDay)
+	if gcRetentionDays < 1 {
+		log.Infof("skipping gc of eth hashes as retention days is less than 1")
+		return
+	}
```

Comment:
oh, so if you set it anything less than 2880 then you're going to have hashes building up? maybe setting it to less than 2880 should be an error on startup if this is necessary, alternatively convert this to minutes and delete more fine-grained?

Reply:
@rvagg Filed https://github.com/filecoin-project/lotus/issues/12465 so we can do it as a follow up. It's a good first issue for a new contributor.

Reply:
If we're deferring support then then I think it should be an error to set it to anything `0>x<2880`, this is going to be a surprise for anyone not info logging and watching their logs

Reply:
@rvagg This is done. Have added better docs for this param and we now throw an error if it's less than 2880 with an informative error message.